File: backend/app/models/loan.py
from sqlalchemy import Column, Integer, String, Decimal, Date, DateTime, ForeignKey
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship
from datetime import datetime

Base = declarative_base()

class Loan(Base):
    __tablename__ = "loans"
    
    id = Column(Integer, primary_key=True, index=True)
    borrower_name = Column(String, index=True)
    co_borrower_name = Column(String)
    address = Column(String)
    city = Column(String)
    state = Column(String)
    zip_code = Column(String)
    loan_amount = Column(Decimal(15, 2))
    interest_rate = Column(Decimal(5, 4))
    maturity_date = Column(Date)
    date_of_loan = Column(Date)
    current_upb = Column(Decimal(15, 2))
    accrued_interest = Column(Decimal(15, 2))
    total_balance = Column(Decimal(15, 2))
    last_paid_date = Column(Date)
    next_due_date = Column(Date)
    remaining_term = Column(Integer)
    legal_status = Column(String)
    lien_position = Column(String)
    investor_name = Column(String)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    documents = relationship("Document", back_populates="loan")

class Document(Base):
    __tablename__ = "documents"
    
    id = Column(Integer, primary_key=True, index=True)
    loan_id = Column(Integer, ForeignKey("loans.id"))
    document_type = Column(String)
    recording_date = Column(Date)
    instrument_number = Column(String)
    original_lender = Column(String)
    assignor = Column(String)
    assignee = Column(String)
    confidence_score = Column(Decimal(3, 2))
    created_at = Column(DateTime, default=datetime.utcnow)
    
    loan = relationship("Loan", back_populates="documents")

File: backend/src/models/Loan.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const Loan = sequelize.define('Loan', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  loan_number: {
    type: DataTypes.STRING(100),
    allowNull: true,
    unique: true
  },
  borrower_name: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  co_borrower_name: {
    type: DataTypes.STRING(255),
    allowNull: true
  },
  property_address: {
    type: DataTypes.STRING(500),
    allowNull: false
  },
  property_city: {
    type: DataTypes.STRING(100),
    allowNull: true
  },
  property_state: {
    type: DataTypes.STRING(50),
    allowNull: true
  },
  property_zip: {
    type: DataTypes.STRING(20),
    allowNull: true
  },
  loan_amount: {
    type: DataTypes.DECIMAL(15, 2),
    allowNull: true
  },
  interest_rate: {
    type: DataTypes.DECIMAL(5, 4),
    allowNull: true,
    validate: {
      min: 0,
      max: 1 // Stored as decimal (e.g., 0.05 for 5%)
    }
  },
  maturity_date: {
    type: DataTypes.DATEONLY,
    allowNull: true
  },
  loan_date: {
    type: DataTypes.DATEONLY,
    allowNull: true
  },
  current_upb: {
    type: DataTypes.DECIMAL(15, 2),
    allowNull: true
  },
  accrued_interest: {
    type: DataTypes.DECIMAL(15, 2),
    allowNull: true
  },
  total_balance: {
    type: DataTypes.DECIMAL(15, 2),
    allowNull: true
  },
  last_paid_date: {
    type: DataTypes.DATEONLY,
    allowNull: true
  },
  next_due_date: {
    type: DataTypes.DATEONLY,
    allowNull: true
  },
  remaining_term: {
    type: DataTypes.INTEGER,
    allowNull: true
  },
  legal_status: {
    type: DataTypes.STRING(100),
    allowNull: true
  },
  lien_position: {
    type: DataTypes.STRING(50),
    allowNull: true
  },
  investor_name: {
    type: DataTypes.STRING(255),
    allowNull: true
  },
  servicer_name: {
    type: DataTypes.STRING(255),
    allowNull: true
  },
  monthly_payment: {
    type: DataTypes.DECIMAL(15, 2),
    allowNull: true
  },
  loan_type: {
    type: DataTypes.STRING(50),
    allowNull: true
  },
  property_type: {
    type: DataTypes.STRING(50),
    allowNull: true
  },
  next_due_date: {
    type: DataTypes.DATEONLY,
    allowNull: true
  },
  document_types: {
    type: DataTypes.STRING(255),
    allowNull: true
  },
  // Future-proofing columns
  servicing_data: {
    type: DataTypes.JSONB,
    allowNull: true
  },
  property_data: {
    type: DataTypes.JSONB,
    allowNull: true
  },
  credit_data: {
    type: DataTypes.JSONB,
    allowNull: true
  },
  analytics_data: {
    type: DataTypes.JSONB,
    allowNull: true
  }
}, {
  tableName: 'loans',
  timestamps: true,
  underscored: true,
  indexes: [
    {
      fields: ['loan_number']
    },
    {
      fields: ['borrower_name']
    },
    {
      fields: ['property_address']
    }
  ]
});

module.exports = Loan;